bokehjs/src/lib/models/mappers/categorical_mapper.ts

```diff
@@ -1,48 +1,43 @@
 import type {Mapper} from "./mapper"
-import type {FactorSeq, Factor, L1Factor, L2Factor, L3Factor} from "../ranges/factor_range"
-import type {Arrayable, ArrayableOf} from "core/types"
+import type {FactorSeq, Factor} from "../ranges/factor_range"
+import type {Arrayable} from "core/types"
 import {index_of, find_index} from "core/util/arrayable"
-import {isString} from "core/util/types"
+import {isArray} from "core/util/types"
 import type * as p from "core/properties"
 
 export function _cat_equals(a: ArrayLike<unknown>, b: ArrayLike<unknown>): boolean {
-  if (a.length != b.length)
+  if (a.length != b.length) {
     return false
+  }
 
-  for (let i = 0, end = a.length; i < end; i++) {
+  const n = a.length
+  for (let i = 0; i < n; i++) {
     if (a[i] !== b[i])
       return false
   }
 
   return true
 }
 
-export function cat_v_compute<T>(data: ArrayableOf<Factor>, factors: Arrayable<Factor>,
+export function cat_v_compute<T>(data: Arrayable<unknown>, factors: Arrayable<Factor>,
     targets: Arrayable<T>, values: Arrayable<T>, start: number, end: number | null, extra_value: T): void {
-  const N = data.length
-
-  for (let i = 0; i < N; i++) {
-    let d = data[i]
-
+  let i = 0
+  for (const datum of data) {
```

Comment:
very obscure word, suggest this instead
```suggestion
  for (const item of data) {
```